<launch>

<arg name="playback" default="false"/>
<arg name="visualizar" default="true"/>
<arg name="acumular" default="false"/>
<arg name="reconfigurar" default="false"/>

<arg name="camera_frame"  default="/duo3d/camera_frame"/>
<arg name="optical_frame" default="duo3d_camera"       />
<arg name="base_frame"    default="base_link2"          />
<arg name="odom_frame"    default="map"               />

<!--lancando o driver da camera-->
<group unless="$(arg playback)">
    <include file="$(find duo3d_driver)/launch/duo3d_depth.launch">
          <arg name="feedback" value="false"/>
    </include>
</group>

<!--republicar com a calibracao correta para o resto do pipeline-->
<node name="republish_info_left" pkg="duo3d_slam_pkg" type="republish_cam_info" output="screen">
    <param name="image_topic_in" value="/duo3d/left/image_rect"/>
    <param name="image_topic_out" value="/duo3d/left/image_rect1"/>
    <param name="cam_info_out" value="/duo3d/left/camera_info1"/>
    <param name="camera_name" value="duo3d_left"/>
    <param name="file" value="file://$(find duo3d_slam_pkg)/calib/duo3d_left.yaml"/>
</node>
<node name="republish_info_right" pkg="duo3d_slam_pkg" type="republish_cam_info" output="screen">
    <param name="image_topic_in" value="/duo3d/right/image_rect"/>
    <param name="image_topic_out" value="/duo3d/right/image_rect1"/>
    <param name="cam_info_out" value="/duo3d/right/camera_info1"/>
    <param name="camera_name" value="duo3d_right"/>
    <param name="file" value="file://$(find duo3d_slam_pkg)/calib/duo3d_right.yaml"/>
</node>

<!--Fazendo nuvem de pontos a partir de imagens na ocasiao de tocando bags-->
<group if="$(arg playback)">
    <arg name="pi/2" value="1.5707963267948966" />
    <node pkg="tf" type="static_transform_publisher" name="duo3d_base_link"   args="0 0 0 0 0 $(arg pi/2) $(arg camera_frame) $(arg optical_frame) 100" />
    <node pkg="tf" type="static_transform_publisher" name="duo3d_base_camera" args="0 0 0 0 0 0 $(arg base_frame) $(arg camera_frame) 100" />
<!--    <node pkg="tf" type="static_transform_publisher" name="map_odom"          args="0 0 0 0 0 0 map $(arg odom_frame) 100" />-->

    <group ns="stereo">
    <node name="stereo_image_proc" pkg="stereo_image_proc" type="stereo_image_proc" output="screen">
        <remap from="/stereo/left/image_raw"    to="/duo3d/left/image_rect1"/>
        <remap from="/stereo/right/image_raw"   to="/duo3d/right/image_rect1"/>
        <remap from="/stereo/left/camera_info"  to="/duo3d/left/camera_info1"/>
        <remap from="/stereo/right/camera_info" to="/duo3d/right/camera_info1"/>

        <param name="approximate_sync" value="true"/>
        <param name="queue_size" value="10"/>
        <!--Melhorar visao stereo -->
        <param name="prefilter_size" value="77"/>
        <param name="prefilter_cap" value="34"/>
        <param name="disparity_range" value="50"/>
        <param name="min_disparity" value="0"/>
        <param name="correlation_window_size" value="35"/>
        <param name="speckle_size" value="1000"/>
        <param name="speckle_range" value="3"/>
        <param name="fullDP" value="true"/>
        <param name="uniqueness_ratio" value="1.0"/>
        <param name="texture_threshold" value="1000"/>
    </node>
    </group>
</group>

<!--Para ajustar visao estereo em caso de bag, com stereo_image_proc-->
<group if="$(arg reconfigurar)">
    <node name="rqt_reconfigure" pkg="rqt_reconfigure" type="rqt_reconfigure" output="screen">
    </node>
</group>

<!--lancando o VISO2 com as configuracoes necessarias-->
<node pkg="viso2_ros" type="stereo_odometer" name="stereo_odometer" output="screen">
    <remap from="stereo" to="/duo3d"/>
    <remap from="image" to="image_rect1"/>
    <remap from="camera_info" to="camera_info1"/>
    <param name="sensor_frame_id" value="$(arg camera_frame)"/>
    <param name="base_link_frame_id" value="$(arg base_frame)"/>
    <param name="approximate_sync" value="false"/>
    <param name="publish_tf" value="true"/>
    <param name="queue_size" value="10"/>
    <param name="odom_frame_id" value="$(arg odom_frame)"/>

    <!--Ajustar melhores valores para fazer o match das nuvens-->
    <param name="max_features" value="2"/>
    <param name="bucket_width" value="50"/>
    <param name="bucket_height" value="=50"/>
    <param name="nms_n" value="3"/>
    <param name="nms_tau" value="50"/>
    <param name="match_binsize" value="50"/>
    <param name="match_radius" value="300"/> <!--esse parametro e bastante importante, afeta a distancia entre features-->
    <param name="outlier_disp_tolerance" value="5"/>
    <param name="outlier_flow_tolerance" value="5"/>
    <param name="match_disp_tolerance" value="5"/>
    <param name="half_resolution" value="1"/> <!--melhor esse mesmo-->
    <param name="refinement" value="1"/>
    <param name="ransac_iters" value="300"/>
    <param name="inlier_threshold" value="2.5"/>
    <param name="reweighting" value="true"/> <!--default ja-->
    <param name="ref_frame_change_method" value="2"/>
    <param name="ref_frame_motion_threshold" value="5.0"/>
    <param name="ref_frame_inlier_threshold" value="150"/>

</node>

<!--aqui lancar a acumulacao a partir de overlap-->
<group if="$(arg acumular)">
    <node name="acumular_viso2" pkg="duo3d_slam_pkg" type="acumular_viso2" output="screen">
        <param name="camera_calibration_yaml"  value="file://$(find duo3d_slam_pkg)/calib/duo3d"/>
        <param name="cloud_topic_in"  value="/stereo/points2"                                   />
        <param name="odom_topic_in"   value="/stereo_odometer/odometry"                         />
        <param name="cloud_topic_out" value="/nuvem_acumulada"                                  />
        <param name="overlap_rate"    value="90"                                                />
    </node>
</group>

<!--visualizar os resultados no rviz-->
<node if="$(arg visualizar)" pkg="rviz" type="rviz" name="rviz_orbslam" args="-d $(find duo3d_slam_pkg)/config/viso2_vis.rviz">
</node>

</launch>
